### DISCUSSION TITLE
Enable private properties when using forceEntityConstructor

### DISCUSSION THREAD

**Discussion by mjroeleveld** [mikro-orm/mikro-orm#4524] [Q&A]
**Is your feature request related to a problem? Please describe.**
When using `forceEntityConstructor`, I would expect that I can make entity properties private. 

Use case: when applying DDD patterns, your domain logic should live on your model objects ([Martin Fowler](https://martinfowler.com/eaaCatalog/domainModel.html)). This means that you invoke methods on models that enforce the invariants. For this reason, I want to make some data properties private, like relational properties (`Collection` type). Only setters can be invoked to add items to a collection, ensuring the domain object's invariants to remain intact.

**Describe the solution you'd like**
Currently, when making a property private, I am getting the following error during populate:

`error TS2322: Type '"X"' is not assignable to type 'keyof Y'.`

Is there any way this could be implemented, or is population only possible after invoking the constructor?

**Comment by B4nan** [comment]
Can you provide some repro for this, I am not really sure what you are trying to do.

With that said, I kinda don't think there is much we can do, this sounds more like a limitation of TS, as private properties are not assignable from outside of the class context, e.g. you cannot query by a private property as its not part of the public interface and is not visible from outside. `forceEntityConstructor` is a runtime option, it forces the ORM to use constructors when creating managed entity instances, which you need to do if you want to have native private properties in your class, but still it has nothing to do with how things work on type level.

**Comment by mjroeleveld** [comment]
A simple example:

```typescript
export class Calendar {
  @OneToMany({
    entity: () => CalendarEvent,
    mappedBy: 'calendar',
  })
  private events: Collection<CalendarEvent>;

  constructor({
    events
  }: {
    events: Collection<CalendarEvent>;
  }) {
    this.events = events;
  }

  public addEvent(event: CalendarEvent) {
    // Some checks...
    this.events.add(event);
  }
}
```

I guess I could add some method `load` that loads the necessary items, and invoke that. Then, I can make the property private?

**Reply by B4nan** [reply]
As said above, this is not really about `forceEntityConstructor`, this is a TS limitation, if you do it like that, you won't be able to query for the collection as TS won't expose any non-public properties on type-level (e.g. in `FilterQuery`). More about that in this issue: https://github.com/microsoft/TypeScript/issues/22677

Also note that the ORM will never give you `events` collection into the constructor, that's not how the hydration works. You should have the initializer defined as in the docs (so `events = new Collection(this)`), otherwise it won't be there.
